Marlin/src/feature/touch/xpt2046.cpp

```diff
@@ -25,9 +25,15 @@
 #include "../../inc/MarlinConfig.h"
 #include "../../lcd/dogm/ultralcd_DOGM.h" // for LCD_FULL_PIXEL_WIDTH, etc.
 
+// Touch screen resolution independent of display resolution
+#define TOUCH_SCREEN_HEIGHT 240
+#define TOUCH_SCREEN_WIDTH 320
+
+// Coordinates in terms of touch area
 #define BUTTON_AREA_TOP 175
 #define BUTTON_AREA_BOT 234
-#define SCREEN_START_TOP ((LCD_PIXEL_OFFSET_Y) * 240 / (LCD_FULL_PIXEL_HEIGHT))
+
+#define SCREEN_START_TOP ((LCD_PIXEL_OFFSET_Y) * (TOUCH_SCREEN_HEIGHT) / (LCD_FULL_PIXEL_HEIGHT))
```

Comment:
To my naive view, this seems like it should be this instead…
```cpp
((LCD_PIXEL_OFFSET_Y) * (LCD_FULL_PIXEL_HEIGHT) / (TOUCH_SCREEN_HEIGHT))
```

Reply:
This code: 
```cpp
  const uint16_t x = uint16_t(((uint32_t(getInTouch(XPT2046_X))) * tsoffsets[0]) >> 16) + tsoffsets[1],
                 y = uint16_t(((uint32_t(getInTouch(XPT2046_Y))) * tsoffsets[2]) >> 16) + tsoffsets[3];
```
Transform the X Y to a 320x240 space. Is because that, the it can find the buttons in these coordinates:
```cpp
    return WITHIN(x,  14,  77) ? EN_D
         : WITHIN(x,  90, 153) ? EN_A
         : WITHIN(x, 166, 229) ? EN_B
         : WITHIN(x, 242, 305) ? EN_C
         : 0;
```
It the x y were in a different coordinate space, it won't find the buttons in the above values.

To, the read_buttons just works in 320x240.

This is the reason I do that:
```cpp
SCREEN_START_TOP ((LCD_PIXEL_OFFSET_Y) * 240 / (LCD_FULL_PIXEL_HEIGHT))
```
Its just a rule of the to find LCD_PIXEL_OFFSET_Y in a 320x240 coordinate space, as the read-buttons expects it.

I know, is a little confusion, but it is. I have tested all the way.

If we use LCD_FULL_PIXEL_* in TOUCH_SCREEN_*, all the code must be upscaled too. But I dont have a 2.4/2.8 screen to test the buttons coordinates, so I kept it fixed as it ever was, and just upscaled the other things around.

This way keeps compatibility, because the default TFT is exactly the same the code expect.

By now is the better way to keep.

I was talking to @tpruvot in separate the code for 320 and 480 screens. But again, it will bring a lot of changes and it would delay a lot the merge.

I already working a 2.0 touchable, with support to double touch, and so, I will rewrite the upscale thing. Right now the point was support 3.5 without minor changes and without breaking marlin to others. 

I can reach me if you want.

Reply:
Most of the time `240 / (LCD_FULL_PIXEL_HEIGHT)` will give back 1, which is just what you want if the `SCREEN_START_TOP` is in an un-scaled 320x240 coordinate space.

However, if the value of `LCD_FULL_PIXEL_HEIGHT` is 480 then you are dividing `LCD_PIXEL_OFFSET_Y` by two. I'm just triple-checking to see that this is what you really want. It seems like you'd want to multiply by two to get into screen-space. But, you would divide by two to go from screen-space to touch-space.

It just seemed backward. If the value of `SCREEN_START_TOP` is always converted into the correct space before it is used, please diregard.

Reply:
LCD_PIXEL_OFFSET_Y 32
LCD_FULL_PIXEL_HEIGHT 480
LCD_FULL_PIXEL_WIDTH 320 

The read buttons expect and work with values in 320x240. But LCD_PIXEL_OFFSET_Y is in the same space as LCD_FULL_PIXEL*. So a need to scale it to 320x240 before use in the read-buttons function.

(LCD_PIXEL_OFFSET_Y) * TOUCH_SCREEN_HEIGHT / (LCD_FULL_PIXEL_HEIGHT))
32 * 240 / 320 = 24 pixel

So, if use 32 pixel y offset in a 480x320, it represents 24 pixel y offset in the 320x240, as the function expects.

I drawn it on paper yesterday  and tested it yesterday in a 3.5" 480x320 screen. I changed LCD_PIXEL_OFFSET_Y to 100, to screen goes a way to bottom. Tested the touch and it really began in the first pixel drawn.
After I changed LCD_PIXEL_OFFSET_Y to 32 default, and tested again. The touch began in the right position too.
Same for X and width.

You can try if you have a 3.5" screen.